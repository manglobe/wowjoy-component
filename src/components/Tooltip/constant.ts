export const ARROW_OFFSET = 15;
export const ARROW_WIDTH = 6;
export const ARROW_HEIGHT = 5;
