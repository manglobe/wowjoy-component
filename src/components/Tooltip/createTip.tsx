import * as React from "react";
import * as ReactDOM from "react-dom";
import ControllSwitchHoc from "../../tools/Hoc/ControllSwitchHoc";
import { placement, formatPlacement, convertToPlacement, reverseDirection } from "./placements";
import { TriggerRect, contentRect } from "./tooltip";

interface RequiredProps {
  visible?: boolean;
  autoAdjustOverflow?: boolean;
  mouseEnterDelay?: number;
  mouseLeaveDelay?: number;
  onVisibleChange?: (visible: boolean) => void;
  placement: placement;
  getContainer?: () => HTMLElement;
}
interface State {
  display: "none" | "block";
  contentRect: contentRect;
  triggerRect: TriggerRect;
  placement: placement;
}

export default <P extends object>(OldCpmponent: React.ComponentType<P>) => {
  class Tooltip extends React.PureComponent<RequiredProps, State> {
    static defaultProps = {
      autoAdjustOverflow: true,
      mouseEnterDelay: 0.1,
      mouseLeaveDelay: 0.1,
      duration: 0.3,
      defaultVisible: false,
      placement: "top" as placement,
      arrowMargin: 4,
      theme: "dark"
    };
    constructor(props: RequiredProps) {
      super(props);
      this.contentRef = React.createRef();
      this.state = {
        display: props.visible ? "block" : "none",
        contentRect: { width: 0, height: 0 },
        triggerRect: { left: 0, top: 0, width: 0, height: 0, scrollX: 0, scrollY: 0 },
        placement: props.placement
      };
    }
    contentRef: React.RefObject<HTMLDivElement>;
    triggerRef: HTMLElement;
    leaveTimer: number;
    enterTimer: number;

    componentDidMount() {
      if (!this.triggerRect || !this.contentRect) return;
      this.setState({
        triggerRect: this.triggerRect,
        contentRect: this.contentRect
      });
    }

    componentWillReceiveProps(nextProps: RequiredProps) {
      if (nextProps.visible !== this.props.visible) {
        if (nextProps.visible) {
          this.setState({ display: "block" }, () => {
            const triggerRect = this.triggerRect;
            if (!triggerRect) return;
            const contentRect = this.contentRect;
            if (!contentRect) return;
            let placement = this.props.placement;
            if (nextProps.autoAdjustOverflow) {
              const directions = formatPlacement(placement);
              const newDrections = reverseDirection(directions, { triggerRect, contentRect });
              placement = convertToPlacement(newDrections);
            }
            this.setState({ triggerRect, contentRect, placement });
          });
        }
      }
    }

    get triggerRect() {
      if (!this.triggerRef) return false;
      const { height, width, left, top } = (ReactDOM.findDOMNode(
        this.triggerRef
      ) as HTMLElement).getBoundingClientRect();
      const scrollX = left + (document.documentElement.scrollLeft || document.body.scrollLeft);
      const scrollY = top + (document.documentElement.scrollTop || document.body.scrollTop);
      return { scrollX, scrollY, left, top, height, width };
    }

    get contentRect() {
      if (!this.contentRef.current) return false;
      const width = this.contentRef.current.offsetWidth;
      const height = this.contentRef.current.offsetHeight;
      // console.log("=====", width, height);
      return { width, height };
    }

    setTriggerRect = () => {
      if (!this.triggerRect) return;
      this.setState({ triggerRect: this.triggerRect });
    };

    setContentRect = () => {
      if (!this.contentRect) return;
      this.setState({ contentRect: this.contentRect });
    };

    saveTriggerRef = (node: any) => {
      this.triggerRef = node;
      const { ref } = this.props.children as { ref: any };
      if (typeof ref === "function") {
        ref(node);
      }
    };

    tipMouseEnter = () => {
      this.leaveTimer && clearTimeout(this.leaveTimer);
    };

    tipMouseLeave = () => {
      this.enterTimer && clearTimeout(this.enterTimer);
      this.props.onVisibleChange && this.props.onVisibleChange(false);
    };

    onAnimationEnd = () => {
      if (!this.contentRef.current) return;
      if (!this.props.visible) {
        this.setState({ display: "none" });
      }
    };

    childrenMouseEnter = () => {
      clearTimeout(this.leaveTimer);
      this.enterTimer = setTimeout(() => {
        this.props.onVisibleChange && this.props.onVisibleChange(true);
        this.enterTimer = null;
      }, this.props.mouseEnterDelay * 1000);
    };

    childrenMouseLeave = () => {
      clearTimeout(this.enterTimer);
      this.leaveTimer = setTimeout(() => {
        this.props.onVisibleChange && this.props.onVisibleChange(false);
        this.leaveTimer = null;
      }, this.props.mouseLeaveDelay * 1000);
    };

    render() {
      const {
        contentRef,
        saveTriggerRef,
        tipMouseEnter,
        tipMouseLeave,
        onAnimationEnd,
        childrenMouseEnter,
        childrenMouseLeave,
        setTriggerRect,
        setContentRect
      } = this;
      const { display, contentRect, triggerRect, placement } = this.state;
      return (
        <OldCpmponent
          {...this.props as P}
          {...{
            contentRef,
            triggerRef: saveTriggerRef,
            tipMouseEnter,
            tipMouseLeave,
            onAnimationEnd,
            childrenMouseEnter,
            childrenMouseLeave,
            contentDisplay: display,
            contentRect,
            triggerRect,
            placement,
            setTriggerRect,
            setContentRect
          }}
        />
      );
    }
  }
  return ControllSwitchHoc({
    value: "visible",
    defaultValue: "defaultVisible",
    onChange: "onVisibleChange"
  })(Tooltip);
};
