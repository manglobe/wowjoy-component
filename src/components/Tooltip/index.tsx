import ComplexTooltip from "./ComplexTooltip";
import SimpleTooltip from "./SimpleTooltip";

export { SimpleTooltip };
export default ComplexTooltip;
