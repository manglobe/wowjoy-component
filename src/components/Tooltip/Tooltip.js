import PropTypes from "prop-types";
import React, { PureComponent } from "react";
import { createPortal, findDOMNode } from "react-dom";
import styled, { css, keyframes } from "styled-components";
import getPlacements from "./placements";
import ControllSwitchHoc from "./ControllSwitchHoc";
import { ARROW_WIDTH, ARROW_HEIGHT } from "./constant";

const enterAnimation = keyframes`
  0%{
    transform: scale(0.8);
    opacity:0;
  }
  100%{
    transform: scale(1);
    opacity: 1;
  }
`;
const leaveAnimation = keyframes`
  0%{
    transform: scale(1);
    opacity:1;
  }
  100%{
    transform: scale(0.8);
    opacity:0;
  }
`;

const TriggerBox = styled.span`
  display: inline-block;
`;
const Layer = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
`;

const Content = styled.div`
  position: absolute;
  max-width: 300px;
  ${p =>
    p.visible &&
    css`
      animation: ${p.enterAnimation || enterAnimation} 0.1s forwards;
    `}
  ${p =>
    !p.visible &&
    css`
      animation: ${p.leaveAnimation || leaveAnimation} 0.1s forwards;
    `}
  ${p => p.defaultStyles}
`;

const Inner = styled.div`
  width: max-content;
  max-width: 100%;
  padding: 6px 8px;
  word-break: break-word;
  min-height: 30px;
  border-radius: 3px;
  ${p =>
    p.theme === "dark"
      ? `background-color: rgba(0, 0, 0, 0.5);box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.3);color: #ffffff;`
      : `background: #FFFFFF;box-shadow: 0 0 4px 0 rgba(0,0,0,0.3);`}
  font-family: "STHeitiSC-Light";
  font-size: 12px;
  letter-spacing: 0;
  line-height: 20px;
  text-align: left;
  display: flex;
  align-items: center;
  ::before {
    content: "";
    display: inline-block;
    width: 0;
    height: 0;
    border-color: transparent;
    border: ${ARROW_WIDTH}px solid transparent;
    ${p =>
      `border-${p.borderDerectionClass}-color: ${
        p.theme === "dark" ? "rgba(0, 0, 0, 0.4)" : "#fff"
      };`}
    ${p => `border-${p.borderDerectionClass}-width: ${ARROW_HEIGHT}px`};
    position: absolute;
    transition: 0.1s;
    ${p => p.arrowStyle}
  }
`;

class Tooltip extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      contentRect: { width: 0, height: 0 },
      triggerRect: { x: 0, y: 0, width: 0, height: 0 }
    };
    this.getPlacement = getPlacements(props.placement, {
      autoAdjustOverflow: props.autoAdjustOverflow,
      arrowPointAtCenter: props.arrowPointAtCenter
    });
    this.setScrollOffset();
  }
  layerRef;
  contentRef;
  innerRef;
  enterTimer;
  leaveTimer;
  triggerNode;
  componentDidMount() {
    if (this.props.visible) {
      this.contentRef.style.display = "block";
      this.setContentRect();
    } else {
      this.contentRef.style.display = "none";
    }
    this.setTriggerRect();
  }

  componentWillReceiveProps(nextProps, nextState) {
    if (nextProps.visible !== this.props.visible) {
      if (nextProps.visible) {
        this.contentRef.style.display = "block";
        this.setContentRect();
      }
      if (this.props.isControlled) {
        this.props.onVisibleChange && this.props.onVisibleChange(nextProps.visible);
      }
    }
    this.setTriggerRect();
  }

  setScrollOffset = () => {
    this.documentScrollTop = document.documentElement.scrollTop || document.body.scrollTop;
    this.documentScrollLeft = document.documentElement.scrollLeft || document.body.scrollLeft;
  };

  setContentRect = () => {
    this.setState({
      contentRect: {
        width: this.contentRef.clientWidth,
        height: this.contentRef.clientHeight
      }
    });
  };

  getTriggerNode() {
    if (!this.triggerNode) {
      this.triggerNode = findDOMNode(this.childrenRef);
    }
    return this.triggerNode;
  }

  setTriggerRect = () => {
    const { x, y, height, width } = this.getTriggerNode().getBoundingClientRect();
    this.setState({
      triggerRect: {
        scrollX: x + this.documentScrollLeft,
        scrollY: y + this.documentScrollTop,
        x,
        y,
        height,
        width
      }
    });
  };

  get currentPlace() {
    return this.getPlacement({
      triggerRect: this.state.triggerRect,
      contentRect: this.state.contentRect
    });
  }

  getPopPositionStyle = contentOffset => {
    if (!this.contentRef) return ``;
    return `
      left: ${contentOffset[0]}px;
      top: ${contentOffset[1]}px;
    `;
  };

  onMouseEnter = e => {
    this.leaveTimer && clearTimeout(this.leaveTimer);
  };
  onMouseLeave = e => {
    this.enterTimer && clearTimeout(this.enterTimer);
    this.props.onVisibleChange && this.props.onVisibleChange(false);
  };
  onAnimationEnd = e => {
    if (!this.props.visible) {
      this.contentRef.style.display = "none";
    }
  };

  onChildrenMouseEnter = e => {
    this.setScrollOffset();
    this.setTriggerRect();
    clearTimeout(this.leaveTimer);
    this.enterTimer = setTimeout(() => {
      this.props.onVisibleChange && this.props.onVisibleChange(true);
      this.enterTimer = null;
    }, this.props.mouseEnterDelay * 1000);
  };
  onChildrenMouseLeave = e => {
    clearTimeout(this.enterTimer);
    this.leaveTimer = setTimeout(() => {
      this.props.onVisibleChange && this.props.onVisibleChange(false);
      this.leaveTimer = null;
    }, this.props.mouseLeaveDelay * 1000);
  };

  getChildren = () => {
    const { children, isControlled } = this.props;
    let triggerProps = isControlled
      ? {}
      : {
          onMouseEnter: this.onChildrenMouseEnter,
          onMouseLeave: this.onChildrenMouseLeave
        };
    if (typeof children === "string" || React.Children.count(children) > 1) {
      return (
        <TriggerBox
          className="wjc-tooltip-triggerbox"
          ref={ref => (this.childrenRef = ref)}
          {...triggerProps}
        >
          {children}
        </TriggerBox>
      );
    }
    return React.cloneElement(this.props.children, {
      ...triggerProps,
      ref: node => {
        this.childrenRef = node;
        const { ref } = this.props.children;
        if (typeof ref === "function") {
          ref(node);
        }
      }
    });
  };

  render() {
    const {
      defaultStyles,
      className,
      title,
      enterAnimation,
      leaveAnimation,
      placement,
      getContainer,
      theme,
      visible
    } = this.props;
    const {
      arrowStyle,
      borderDerectionClass,
      popBoxStyle,
      transformOrigin,
      contentOffset
    } = this.currentPlace;

    return (
      <React.Fragment>
        {this.getChildren()}
        {createPortal(
          <Layer
            ref={ref => (this.layerRef = ref)}
            defaultStyles={defaultStyles}
            className={`wjc-tooltip-layer ${placement ? `wjc-tooltip-${placement}` : ""} ${
              className ? className : ""
            }`}
          >
            <Content
              ref={ref => (this.contentRef = ref)}
              className={`wjc-tooltip-conent`}
              defaultStyles={`
                  ${this.getPopPositionStyle(contentOffset)}
                  ${popBoxStyle}
                  transform-origin: ${transformOrigin};
                `}
              enterAnimation={enterAnimation}
              leaveAnimation={leaveAnimation}
              visible={visible}
              onMouseEnter={this.onMouseEnter}
              onMouseLeave={this.onMouseLeave}
              onAnimationEnd={this.onAnimationEnd}
            >
              <Inner
                ref={ref => (this.innerRef = ref)}
                className={`wjc-tooltip-inner`}
                arrowStyle={arrowStyle}
                borderDerectionClass={borderDerectionClass}
                theme={theme}
              >
                {title}
              </Inner>
            </Content>
          </Layer>,
          getContainer ? getContainer() : document.body
        )}
      </React.Fragment>
    );
  }
}

Tooltip.defaultProps = {
  arrowPointAtCenter: false,
  autoAdjustOverflow: true,
  defaultVisible: false,
  shouldFirstAnimation: false,
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1,
  placement: "top",
  theme: "dark"
};

Tooltip.propTypes = {
  className: PropTypes.string,
  defaultStyles: PropTypes.string,
  mouseEnterDelay: PropTypes.number,
  mouseLeaveDelay: PropTypes.number,
  placement: PropTypes.oneOf([
    "top",
    "left",
    "right",
    "bottom",
    "topLeft",
    "leftTop",
    "rightTop",
    "topRight",
    "leftBottom",
    "bottomLeft",
    "bottomRight",
    "rightBottom"
  ]),
  enterAnimation: PropTypes.object,
  leaveAnimation: PropTypes.object,
  defaultVisible: PropTypes.bool,
  getContainer: PropTypes.func,
  onVisibleChange: PropTypes.func,
  shouldFirstAnimation: PropTypes.bool,
  theme: PropTypes.oneOf(["dark", "light"])
};
export default ControllSwitchHoc({
  value: "visible",
  defaultValue: "defaultVisible",
  onChange: "onVisibleChange"
})(Tooltip);
