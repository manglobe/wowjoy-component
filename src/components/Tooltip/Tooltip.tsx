import * as React from "react";
import { createPortal, findDOMNode } from "react-dom";
import styled, { css, keyframes, Keyframes } from "styled-components";
import getPlacements, { placement, getPlacement } from "./placements";
import ControllSwitchHoc from "../../tools/Hoc/ControllSwitchHoc";
import { ARROW_WIDTH, ARROW_HEIGHT } from "./constant";

const enterAnimation = keyframes`
  0%{
    transform: scale(0.8);
    opacity:0;
  }
  100%{
    transform: scale(1);
    opacity: 1;
  }
`;
const leaveAnimation = keyframes`
  0%{
    transform: scale(1);
    opacity:1;
  }
  100%{
    transform: scale(0.8);
    opacity:0;
  }
`;

const TriggerBox = styled.span<{ className: string }>`
  display: inline-block;
`;
const Layer = styled.div<{ defaultStyles: string; className: string }>`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
`;

const Content = styled.div<{
  visible: boolean;
  enterAnimation: Keyframes;
  leaveAnimation: Keyframes;
  defaultStyles: string;
}>`
  position: absolute;
  max-width: 300px;
  ${p =>
    p.visible &&
    css`
      animation: ${p.enterAnimation || enterAnimation} 0.1s forwards;
    `}
  ${p =>
    !p.visible &&
    css`
      animation: ${p.leaveAnimation || leaveAnimation} 0.1s forwards;
    `}
  ${p => p.defaultStyles}
`;

const Inner = styled.div<{
  theme: string;
  borderDerectionClass: string;
  arrowStyle: string;
}>`
  width: max-content;
  max-width: 100%;
  padding: 6px 8px;
  word-break: break-word;
  min-height: 30px;
  border-radius: 3px;
  ${p =>
    p.theme === "dark"
      ? `background-color: rgba(0, 0, 0, 0.5);box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.3);color: #ffffff;`
      : `background: #FFFFFF;box-shadow: 0 0 4px 0 rgba(0,0,0,0.3);`}
  font-family: "STHeitiSC-Light";
  font-size: 12px;
  letter-spacing: 0;
  line-height: 20px;
  text-align: left;
  display: flex;
  align-items: center;
  ::before {
    content: "";
    display: inline-block;
    width: 0;
    height: 0;
    border-color: transparent;
    border: ${ARROW_WIDTH}px solid transparent;
    ${p =>
      `border-${p.borderDerectionClass}-color: ${
        p.theme === "dark" ? "rgba(0, 0, 0, 0.4)" : "#fff"
      };`}
    ${p => `border-${p.borderDerectionClass}-width: ${ARROW_HEIGHT}px`};
    position: absolute;
    transition: 0.1s;
    ${p => p.arrowStyle}
  }
`;

interface Props {
  title: React.ReactNode;
  children: React.ReactNode;
  defaultStyles?: string;
  className?: string;
  enterAnimation?: Keyframes;
  leaveAnimation?: Keyframes;
  placement?: placement;
  getContainer?: () => HTMLElement;
  theme?: string;
  visible?: boolean;
  autoAdjustOverflow?: boolean;
  arrowPointAtCenter?: boolean;
  mouseEnterDelay?: number;
  mouseLeaveDelay?: number;
  onVisibleChange?: (visible: boolean) => void;
}
interface State {
  contentRect: { width: number; height: number };
  triggerRect: {
    width: number;
    height: number;
    x: number;
    y: number;
    scrollX: number;
    scrollY: number;
  };
}

class Tooltip extends React.PureComponent<Props, State> {
  static defaultProps = {
    arrowPointAtCenter: false,
    autoAdjustOverflow: true,
    defaultVisible: false,
    shouldFirstAnimation: false,
    mouseEnterDelay: 0.1,
    mouseLeaveDelay: 0.1,
    placement: "top" as placement,
    theme: "dark"
  };
  constructor(props: Props) {
    super(props);
    this.state = {
      contentRect: { width: 0, height: 0 },
      triggerRect: { x: 0, y: 0, width: 0, height: 0, scrollX: 0, scrollY: 0 }
    };
    this.getPlacement = getPlacements(props.placement, {
      autoAdjustOverflow: props.autoAdjustOverflow,
      arrowPointAtCenter: props.arrowPointAtCenter
    });
    this.setScrollOffset();
  }
  getPlacement: getPlacement;
  childrenRef: HTMLElement;
  layerRef: HTMLElement;
  contentRef: HTMLElement;
  innerRef: HTMLElement;
  enterTimer: number;
  leaveTimer: number;
  triggerNode: HTMLElement | Element | Text;
  documentScrollTop: number;
  documentScrollLeft: number;
  componentDidMount() {
    if (this.props.visible) {
      this.contentRef.style.display = "block";
      this.setContentRect();
    } else {
      this.contentRef.style.display = "none";
    }
    this.setTriggerRect();
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.visible !== this.props.visible) {
      if (nextProps.visible) {
        this.contentRef.style.display = "block";
        this.setContentRect();
      }
    }
  }
  did = false;
  componentDidUpdate() {
    if (!this.did) {
      this.did = true;
      this.setTriggerRect();
    } else {
      this.did = false;
    }
  }

  setScrollOffset = () => {
    this.documentScrollTop = document.documentElement.scrollTop || document.body.scrollTop;
    this.documentScrollLeft = document.documentElement.scrollLeft || document.body.scrollLeft;
  };

  setContentRect = () => {
    this.setState({
      contentRect: {
        width: this.contentRef.clientWidth,
        height: this.contentRef.clientHeight
      }
    });
  };

  getTriggerNode() {
    if (!this.triggerNode) {
      this.triggerNode = findDOMNode(this.childrenRef);
    }
    return this.triggerNode;
  }

  setTriggerRect = () => {
    const { x, y, height, width } = (this.getTriggerNode() as any).getBoundingClientRect();
    this.setState({
      triggerRect: {
        scrollX: x + this.documentScrollLeft,
        scrollY: y + this.documentScrollTop,
        x,
        y,
        height,
        width
      }
    });
  };

  get currentPlace() {
    return this.getPlacement({
      triggerRect: this.state.triggerRect,
      contentRect: this.state.contentRect
    });
  }

  getPopPositionStyle = (contentOffset: number[]) => {
    if (!this.contentRef) return ``;
    return `
      left: ${contentOffset[0]}px;
      top: ${contentOffset[1]}px;
    `;
  };

  onMouseEnter = () => {
    this.leaveTimer && clearTimeout(this.leaveTimer);
  };
  onMouseLeave = () => {
    this.enterTimer && clearTimeout(this.enterTimer);
    this.props.onVisibleChange && this.props.onVisibleChange(false);
  };
  onAnimationEnd = () => {
    if (!this.props.visible) {
      this.contentRef.style.display = "none";
    }
  };

  onChildrenMouseEnter = () => {
    this.setScrollOffset();
    this.setTriggerRect();
    clearTimeout(this.leaveTimer);
    this.enterTimer = setTimeout(() => {
      this.props.onVisibleChange && this.props.onVisibleChange(true);
      this.enterTimer = null;
    }, this.props.mouseEnterDelay * 1000);
  };
  onChildrenMouseLeave = () => {
    clearTimeout(this.enterTimer);
    this.leaveTimer = setTimeout(() => {
      this.props.onVisibleChange && this.props.onVisibleChange(false);
      this.leaveTimer = null;
    }, this.props.mouseLeaveDelay * 1000);
  };

  getChildren = () => {
    const { children } = this.props;
    let triggerProps = {
      onMouseEnter: this.onChildrenMouseEnter,
      onMouseLeave: this.onChildrenMouseLeave
    };
    if (typeof children === "string" || React.Children.count(children) > 1) {
      return (
        <TriggerBox
          className="wjc-tooltip-triggerbox"
          ref={ref => (this.childrenRef = ref)}
          {...triggerProps}
        >
          {children}
        </TriggerBox>
      );
    }
    // @ts-ignore
    return React.cloneElement(children, {
      ...triggerProps,
      ref: (node: any) => {
        this.childrenRef = node;
        // @ts-ignore
        const { ref } = children;
        if (typeof ref === "function") {
          ref(node);
        }
      }
    });
  };

  render() {
    const {
      defaultStyles,
      className,
      title,
      enterAnimation,
      leaveAnimation,
      placement,
      getContainer,
      theme,
      visible
    } = this.props;
    const {
      arrowStyle,
      borderDerectionClass,
      popBoxStyle,
      transformOrigin,
      contentOffset
    } = this.currentPlace;
    return (
      <React.Fragment>
        {this.getChildren()}
        {createPortal(
          <Layer
            ref={ref => (this.layerRef = ref)}
            defaultStyles={defaultStyles}
            className={`wjc-tooltip-layer ${placement ? `wjc-tooltip-${placement}` : ""} ${
              className ? className : ""
            }`}
          >
            <Content
              ref={ref => (this.contentRef = ref)}
              className={`wjc-tooltip-conent`}
              defaultStyles={`
                ${this.getPopPositionStyle(contentOffset)}
                ${popBoxStyle}
                transform-origin: ${transformOrigin};
              `}
              enterAnimation={enterAnimation}
              leaveAnimation={leaveAnimation}
              visible={visible}
              onMouseEnter={this.onMouseEnter}
              onMouseLeave={this.onMouseLeave}
              onAnimationEnd={this.onAnimationEnd}
            >
              <Inner
                ref={ref => (this.innerRef = ref)}
                className={`wjc-tooltip-inner`}
                arrowStyle={arrowStyle}
                borderDerectionClass={borderDerectionClass}
                theme={theme}
              >
                {title}
              </Inner>
            </Content>
          </Layer>,
          getContainer ? getContainer() : document.body
        )}
      </React.Fragment>
    );
  }
}

export default ControllSwitchHoc({
  value: "visible",
  defaultValue: "defaultVisible",
  onChange: "onVisibleChange"
})(Tooltip);
