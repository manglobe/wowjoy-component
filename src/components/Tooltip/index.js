import Tooltip from './Tooltip';
import Hoc from '../../tools/Hoc';
import type1 from "./types/type1";

export const Type1 = Hoc(type1)(Tooltip);
export default Tooltip