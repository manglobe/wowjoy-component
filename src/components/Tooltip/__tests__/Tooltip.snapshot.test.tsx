import * as React from "react";
import Tooltip, { SimpleTooltip } from "../index";
import { mount, configure } from "enzyme";
import * as adapter from "enzyme-adapter-react-16";
configure({ adapter: new (adapter as any).default() });

describe("Tooltip", () => {
  const complexTooltip = <Tooltip>Tooltip Content</Tooltip>;
  const simpleTooltip = <SimpleTooltip>Tooltip Content</SimpleTooltip>;
  it("renders correctly", () => {
    const tree1 = mount(complexTooltip).render();
    expect(tree1).toMatchSnapshot();
    const tree2 = mount(simpleTooltip).render();
    expect(tree2).toMatchSnapshot();
  });
});
