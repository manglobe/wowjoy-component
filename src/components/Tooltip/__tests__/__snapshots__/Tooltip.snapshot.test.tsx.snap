// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "wjc-tooltip-triggerbox sc-bdVaJa bSrzPW",
    },
    "children": Array [
      Object {
        "data": "Tooltip Content",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": Object {
      "attribs": Object {
        "class": "wjc-tooltip-layer wjc-tooltip-top  sc-bwzfXH dIozFi",
      },
      "children": Array [
        Object {
          "attribs": Object {
            "class": "wjc-tooltip-conent sc-htpNat gJHZri",
            "style": "display: none;",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "wjc-tooltip-inner sc-bxivhb hnUnaq",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
      ],
      "name": "div",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": [Circular],
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
      },
    },
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
        Object {
          "attribs": Object {
            "class": "wjc-tooltip-layer wjc-tooltip-top  sc-bwzfXH dIozFi",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "wjc-tooltip-conent sc-htpNat gJHZri",
                "style": "display: none;",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "wjc-tooltip-inner sc-bxivhb hnUnaq",
                  },
                  "children": Array [],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "style": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": null,
          "prev": [Circular],
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "1": Object {
    "attribs": Object {
      "class": "wjc-tooltip-layer wjc-tooltip-top  sc-bwzfXH dIozFi",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "wjc-tooltip-conent sc-htpNat gJHZri",
          "style": "display: none;",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "wjc-tooltip-inner sc-bxivhb hnUnaq",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": Object {
      "attribs": Object {
        "class": "wjc-tooltip-triggerbox sc-bdVaJa bSrzPW",
      },
      "children": Array [
        Object {
          "data": "Tooltip Content",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
      ],
      "name": "span",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": [Circular],
      "parent": null,
      "prev": null,
      "root": Object {
        "attribs": Object {},
        "children": Array [
          [Circular],
          [Circular],
        ],
        "name": "root",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": null,
        "prev": null,
        "type": "root",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      "type": "tag",
      "x-attribsNamespace": Object {
        "class": undefined,
      },
      "x-attribsPrefix": Object {
        "class": undefined,
      },
    },
    "root": Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "attribs": Object {
            "class": "wjc-tooltip-triggerbox sc-bdVaJa bSrzPW",
          },
          "children": Array [
            Object {
              "data": "Tooltip Content",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": null,
          "prev": null,
          "root": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 2,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
